Completion error: completion--some: Internal error: ChatHistoryEntry doesn’t match \`E

**Describe the bug**

Using *xref* as the completion provider (not company, not help, just plain built-in xref) with gopls as the server.

About half the time, M-Tab fails with an error such as the following:
```
completion--some: Internal error: ChatHistoryEntry doesn’t match \`E
```

**To Reproduce**
Install lsp-mode and gopls.  Do not install company-mode or helm.  Start M-Tabbing around until you get the error.

**Expected behavior**
Completion works without error.

**Which Language Server did you use**

gopls

**OS**
Debian Linux.

**Error callstack**

```
Debugger entered--Lisp error: (error #("Internal error: ChatHistoryEntry doesn’t match \\`E" 16 32 (lsp-completion-prefix "E" lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-start-point 2467 lsp-sort-text "00003" lsp-completion-item #<hash-table equal 8/10 0x1586c475bf19>) 49 50 (fontified t)))
  signal(error (#("Internal error: ChatHistoryEntry doesn’t match \\`E" 16 32 (lsp-completion-prefix "E" lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-start-point 2467 lsp-sort-text "00003" lsp-completion-item #<hash-table equal 8/10 0x1586c475bf19>) 49 50 (fontified t))))
  error("Internal error: %s doesn't match %s" #("ChatHistoryEntry" 0 16 (lsp-completion-item #<hash-table equal 8/10 0x1586c475bf19> lsp-sort-text "00003" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("\\`E" 2 3 (fontified t)))
  completion-pcm--merge-completions((#("Expire" 0 6 (lsp-completion-item #<hash-table equal 9/12 0x1586c47ace01> lsp-sort-text "00000" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ErrNotAuthorised.Error" 0 22 (lsp-completion-item #<hash-table equal 8/10 0x1586c476f19d> lsp-sort-text "00001" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ErrNotAuthorised" 0 16 (lsp-completion-item #<hash-table equal 8/10 0x1586c476f365> lsp-sort-text "00002" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ChatHistoryEntry" 0 16 (lsp-completion-item #<hash-table equal 8/10 0x1586c475bf19> lsp-sort-text "00003" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("KickError" 0 9 (lsp-completion-item #<hash-table equal 8/10 0x1586c4807235> lsp-sort-text "00004" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ProtocolError" 0 13 (lsp-completion-item #<hash-table equal 8/10 0x1586c47e2ff5> lsp-sort-text "00005" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("UserError" 0 9 (lsp-completion-item #<hash-table equal 8/10 0x1586c47e31bd> lsp-sort-text "00006" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E"))) (#("E" 0 1 (fontified t))))
  completion-pcm--merge-try((#("E" 0 1 (fontified t))) (#("Expire" 0 6 (lsp-completion-item #<hash-table equal 9/12 0x1586c47ace01> lsp-sort-text "00000" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ErrNotAuthorised.Error" 0 22 (lsp-completion-item #<hash-table equal 8/10 0x1586c476f19d> lsp-sort-text "00001" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ErrNotAuthorised" 0 16 (lsp-completion-item #<hash-table equal 8/10 0x1586c476f365> lsp-sort-text "00002" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ChatHistoryEntry" 0 16 (lsp-completion-item #<hash-table equal 8/10 0x1586c475bf19> lsp-sort-text "00003" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("KickError" 0 9 (lsp-completion-item #<hash-table equal 8/10 0x1586c4807235> lsp-sort-text "00004" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("ProtocolError" 0 13 (lsp-completion-item #<hash-table equal 8/10 0x1586c47e2ff5> lsp-sort-text "00005" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E")) #("UserError" 0 9 (lsp-completion-item #<hash-table equal 8/10 0x1586c47e31bd> lsp-sort-text "00006" lsp-completion-start-point 2467 lsp-completion-markers (2467 #<marker (moves after insertion) at 2468 in galene.go>) lsp-completion-prefix "E"))) "" "")
  completion-pcm-try-completion(#("E" 0 1 (fontified t)) #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil 1)
  #f(compiled-function (style) #<bytecode 0x1586c4802d59>)(partial-completion)
  completion--some(#f(compiled-function (style) #<bytecode 0x1586c4802d59>) (basic partial-completion emacs22))
  completion--nth-completion(1 #("E" 0 1 (fontified t)) #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil 1 (metadata (category . lsp-capf) (display-sort-function . identity)))
  completion-try-completion(#("E" 0 1 (fontified t)) #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil 1 (metadata (category . lsp-capf) (display-sort-function . identity)))
  completion--do-completion(#<marker at 2467 in galene.go> 2468)
  completion--in-region-1(#<marker at 2467 in galene.go> 2468)
  #f(compiled-function (start end collection predicate) #<bytecode 0x1ff94db3f12f>)(#<marker at 2467 in galene.go> 2468 #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil)
  apply(#f(compiled-function (start end collection predicate) #<bytecode 0x1ff94db3f12f>) (#<marker at 2467 in galene.go> 2468 #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil))
  #f(compiled-function (funs global args) #<bytecode 0x1586c4802d25>)(nil nil (#<marker at 2467 in galene.go> 2468 #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil))
  completion--in-region(#<marker at 2467 in galene.go> 2468 #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil)
  completion-in-region(#<marker at 2467 in galene.go> 2468 #f(compiled-function (probe pred action) #<bytecode 0x1586c47b4b6d>) nil)
  completion-at-point()
  complete-symbol(nil)
  funcall-interactively(complete-symbol nil)
  call-interactively(complete-symbol nil nil)
  command-execute(complete-symbol)
```

```
2020/12/13 21:09:33 imports fixes: AllImportsFixes: /home/jch/go/src/galene/galene.go:107:17: function must be invoked in go statement
	file="/home/jch/go/src/galene/galene.go"

2020/12/13 21:09:34 imports fixes: AllImportsFixes: /home/jch/go/src/galene/galene.go:107:19: function must be invoked in go statement
	file="/home/jch/go/src/galene/galene.go"

2020/12/13 21:09:35 no signature help: cannot find an enclosing function
	position={106 20}

2020/12/13 21:09:35 no signature help: cannot find an enclosing function
	position={107 19}

2020/12/13 21:09:42 background imports cache refresh starting

2020/12/13 21:09:42 background refresh finished after 69.046275ms

2020/12/13 21:10:13 background imports cache refresh starting

2020/12/13 21:10:13 background refresh finished after 47.584495ms
```

--- a/lsp-completion.el
+++ b/lsp-completion.el
@@ -470,12 +470,14 @@ The MARKERS and PREFIX value will be attached to each candidate."
           ;; boundaries
           ((equal (car-safe action) 'boundaries) nil)
           ;; try-completion
-          ((null action) (cl-first (member probe (funcall candidates))))
+          ((null action)
+           (when-let ((cands (funcall candidates)))
+             (if (cl-rest cands) probe (cl-first cands))))
           ;; test-completion: not return exact match so that the selection will
           ;; always be shown
           ((equal action 'lambda) nil)
           ;; retrieve candidates
-          (t (funcall candidates))))
+          ((equal action t) (funcall candidates))))
        :annotation-function #'lsp-completion--annotate
        :company-candidate-kind #'lsp-completion--candidate-kind
        :company-require-match 'never

